pkg/venafi/tpp/connector.go

```diff
@@ -1231,31 +1265,51 @@ func (c *Connector) getIdentity(userName string) (*policy.IdentityEntry, error)
 		return nil, err
 	}
 
-	return c.getIdentityMatching(resp.Identities, userName)
-}
+	// When TPP looks for a username that matches the filter, an implicit
+	// wildcard is added to the end of the filter string. For example, imagining
+	// that `jsmith` and `jsmithson` are existing identities, searching for
+	// `jsmith` will return both `jsmith` and `jsmithson`. In the case of local
+	// identities, `jsmith` will always be returned first. But in the case of AD
+	// and LDAP, the order of these results may be different, and `jsmithson`
+	// may be unexpectedly returned first. This same problem may appear when an
+	// AD or LDAP provider has been configured to access the local identities:
+	// `jsmithson` may get returned first if `jsmithson` only exists in AD
+	// (because the AD results are returned before the local identities).
+	//
+	// The wildcard problem only affects usernames, not emails. That's because
+	// the LDAP query recommended for enabling user search by email in the
+	// Venafi Configuration Console is based on exact match, unlike `anr` used
+	// for searching usernames. Thus, we do not need to check for an exact match
+	// when an email is provided.
 
-func (c *Connector) getIdentityMatching(identities []policy.IdentityEntry, identityName string) (*policy.IdentityEntry, error) {
-	var identityEntryMatching *policy.IdentityEntry
+	_, err = mail.ParseAddress(filter)
+	isEmail := err == nil
 
-	if len(identities) > 0 {
-		for i := range identities {
-			identityEntry := identities[i]
-			if identityEntry.Name == identityName {
-				identityEntryMatching = &identityEntry
-				break
+	switch {
+	case len(resp.Identities) == 0:
+		return nil, fmt.Errorf("no identity found for '%s'", filter)
+	case len(resp.Identities) == 1:
+		return &resp.Identities[0], nil
+	case len(resp.Identities) > 1 && !isEmail:
+		// The username case: we need to ignore the results that are prefixes of
+		// the queried username. For example, if the filter is `jsmith`, we
+		// ignore `jsmithson` and `jsmithers`.
+		for _, identity := range resp.Identities {
+			if identity.Name == filter {
+				return &identity, nil
 			}
 		}
+		return nil, fmt.Errorf("%d identities were found but none of these identities has a username equal to '%s'", len(resp.Identities), filter)
+	case len(resp.Identities) > 1 && isEmail:
+		// The email case: we do not need to filter out anything. So let's
+		// arbitrarily return the first identity.
+		return &resp.Identities[0], nil
 	}
 
-	//if the identity is not null
-	if identityEntryMatching != nil {
-		return identityEntryMatching, nil
-	} else {
-		return nil, fmt.Errorf("it was not possible to find the user %s", identityName)
-	}
+	return nil, fmt.Errorf("programmer mistake: impossible case in getIdentity")
```

Comment:
Please change this line for the following or something similar 
`return nil, fmt.Errorf("it was not possible to find an Identity using filter value %s", filter)`

Reply:
I suggest prefixing this error with "programmer mistake" because this error isn't meant to happen. If this error occurs, it means that I made a mistake when programming this. "programmer mistake" tells the user that it isn't their fault.

The message "it was not possible to find an Identity using this filter" doesn't describe what has happened. Here, the filter may have returned identities, but none of the switch cases have matched, which is unexpected since these 4 switch cases are meant to catch 100% of the cases. 

I propose the following error; it should help us (developers) debug the issue if it ever happen:

```
programmer mistake: browseIdentities returned 2 results for filter 'jsmith' and none of the switch cases matched
```

Note that panicking would be a good fit here, but panicking isn't a great experience for the callers of VCert, so let's not panic.